### PR TITLE
Tlportal 295 TLPORTAL-296, TLPORTAL-313

### PR DESCRIPTION
**Pull request by dlhaines**

### CODE REVIEW

**Comment by lsloan** [comment]
I didn't see any glaring problems, so I guess it's pretty good.

:shipit:

File: server/ctools_direct_response.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+  def extractAssignment(assignment)
+
+    # create a single assignment object in known format for UI.
+    assign = Hash.new()
+
+    assign[:title] = assignment['calendarItem']['title']
+    assign[:grade] = ''
+    assign[:done] = ''
+    assign[:context] = assignment['calendarItem']['context']['contextTitle']
```
Comment (zqian) [inline server/ctools_direct_response.rb]: any chance the chained value could be null?

File: server/ctools_direct_response.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    collection.map { |dash_entry| extractAssignment(dash_entry) }
+  end
+
+  def extractAssignment(assignment)
+
+    # create a single assignment object in known format for UI.
+    assign = Hash.new()
+
+    assign[:title] = assignment['calendarItem']['title']
+    assign[:grade] = ''
```
Comment (zqian) [inline server/ctools_direct_response.rb]: if the "grade" and "done" always have empty String value, why do we need those two attributes?

Comment (dlhaines) [inline server/ctools_direct_response.rb]: This code is translating from the CTools format into the format expected by the UI.  Those values are required by the UI format but we don't, at this moment, have a translation between CTools and UI formats defined.  It's expected they will get values later.

File: server/ctools_direct_response.rb (hunk 1 above)
Comment (dlhaines) [inline server/ctools_direct_response.rb]: Added checks to return nil if the access path to required data is invalid.  The actual value of data may be nil, but the path can't be nil.

**Pull request merged**